native/dlaqr2.go

```diff
@@ -0,0 +1,382 @@
+				// Dtrexc can not fail in this case.
+				_, ilst, _ = impl.Dtrexc(lapack.EigDecomp, jw, t, ldt, v, ldv, ns-1, ilst, work)
+				ilst++
+			}
+			continue
+		}
+		// Complex conjugate pair.
+		abst := math.Abs(t[(ns-1)*ldt+ns-1]) + math.Sqrt(math.Abs(t[(ns-1)*ldt+ns-2]))*math.Sqrt(math.Abs(t[(ns-2)*ldt+ns-1]))
+		if abst == 0 {
+			abst = math.Abs(s)
+		}
+		if math.Max(math.Abs(s*v[ns-1]), math.Abs(s*v[ns-2])) <= math.Max(smlnum, ulp*abst) {
+			// Deflatable.
+			ns -= 2
+		} else {
+			// Undeflatable, move them up out of the way.
+			// Dtrexc does the right thing with ilst in case of a
+			// rare exchange failure.
+			_, ilst, _ = impl.Dtrexc(lapack.EigDecomp, jw, t, ldt, v, ldv, ns-1, ilst, work)
+			ilst += 2
+		}
+	}
+
+	// Return to Hessenberg form.
+	if ns == 0 {
+		s = 0
+	}
+	if ns < jw {
+		// Sorting diagonal blocks of T improves accuracy for graded
+		// matrices. Bubble sort deals well with exchange failures.
+		sorted := false
+		i := ns
+		for !sorted {
+			sorted = true
+			kend := i - 1
+			i = infqr
+			var k int
+			if i == ns-1 || t[(i+1)*ldt+i] == 0 {
+				k = i + 1
+			} else {
+				k = i + 2
+			}
+			for k <= kend {
+				var evi float64
+				if k == i+1 {
+					evi = math.Abs(t[i*ldt+i])
+				} else {
+					evi = math.Abs(t[i*ldt+i]) + math.Sqrt(math.Abs(t[(i+1)*ldt+i]))*math.Sqrt(math.Abs(t[i*ldt+i+1]))
+				}
+
+				var evk float64
+				if k == kend || t[(k+1)*ldt+k] == 0 {
+					evk = math.Abs(t[k*ldt+k])
+				} else {
+					evk = math.Abs(t[k*ldt+k]) + math.Sqrt(math.Abs(t[(k+1)*ldt+k]))*math.Sqrt(math.Abs(t[k*ldt+k+1]))
+				}
+
+				if evi >= evk {
+					i = k
+				} else {
+					sorted = false
+					_, ilst, ok := impl.Dtrexc(lapack.EigDecomp, jw, t, ldt, v, ldv, i, k, work)
+					if ok {
+						i = ilst
+					} else {
+						i = k
+					}
+				}
+				if i == kend || t[(i+1)*ldt+i] == 0 {
+					k = i + 1
+				} else {
+					k = i + 2
+				}
+			}
+		}
+	}
+
+	// Restore shift/eigenvalue array from T.
+	for i := jw - 1; i >= infqr; {
+		if i == infqr || t[i*ldt+i-1] == 0 {
+			sr[kwtop+i] = t[i*ldt+i]
+			si[kwtop+i] = 0
+			i--
+			continue
+		}
+		aa := t[(i-1)*ldt+i-1]
+		bb := t[(i-1)*ldt+i]
+		cc := t[i*ldt+i-1]
+		dd := t[i*ldt+i]
+		_, _, _, _, sr[kwtop+i-1], si[kwtop+i-1], sr[kwtop+i], si[kwtop+i], _, _ = impl.Dlanv2(aa, bb, cc, dd)
+		i -= 2
+	}
+
+	if ns < jw || s == 0 {
+		if ns > 1 && s != 0 {
+			// Reflect spike back into lower triangle.
+			bi.Dcopy(ns, v[:ns], 1, work[:ns], 1)
+			_, tau := impl.Dlarfg(ns, work[0], work[1:ns], 1)
+			work[0] = 1
+			impl.Dlaset(blas.Lower, jw-2, jw-2, 0, 0, t[2*ldt:], ldt)
+			impl.Dlarf(blas.Left, ns, jw, work[:ns], 1, tau, t, ldt, work[jw:])
+			impl.Dlarf(blas.Right, ns, ns, work[:ns], 1, tau, t, ldt, work[jw:])
+			impl.Dlarf(blas.Right, jw, ns, work[:ns], 1, tau, v, ldv, work[jw:])
+			impl.Dgehrd(jw, 0, ns-1, t, ldt, work[:jw-1], work[jw:], lwork-jw)
+		}
+
+		// Copy updated reduced window into place.
+		if kwtop >= 1 {
```

Comment:
I wanted to include the border value of kwtop in the condition but with `> 0` it seems better. Done.